lib/util/itinerary.js

```diff
@@ -245,31 +245,133 @@ export function getLegBounds (leg) {
   return latLngBounds(coords)
 }
 
-export function routeComparator (a, b) {
-  let aComp, bComp
-  // If route sort order field is available, default sort to these predefined
-  // values. NOTE: OTP/OBA defaults sort order values to -999, so instead of
-  // checking that the sort order is not null, we just check that the values are
-  // not equal. Technically, there could be cases where route_sort_order values
-  // are equivalent, but ideally they should be unique.
-  if (a.sortOrder !== b.sortOrder) {
-    aComp = a.sortOrder
-    bComp = b.sortOrder
-  } else if (!isNaN(parseInt(a.shortName)) && !isNaN(parseInt(b.shortName))) {
-    // Otherwise, if both short names can be parsed as integers, use these
-    // numbers for sorting.
-    aComp = parseInt(a.shortName)
-    bComp = parseInt(b.shortName)
-  } else {
-    // Otherwise, default to short or long name strings.
-    aComp = a.shortName || a.longName
-    bComp = b.shortName || b.longName
+/**
+ * Checks whether an appropriate comparison of numeric values can be made for
+ * sorting purposes. If both values are not valid numbers according to the
+ * isNaN check, then this function returns undefined which indicates that a
+ * secondary sorting criteria should be used instead. If one value is valid and
+ * the other is not, then the valid value will be given sorting priority. If
+ * both values are valid numbers, the difference is obtained as the sort value.
+ *
+ * An optional argument can be provided which will be used to obtain the
+ * comparison value from the comparison function arguments.
+ *
+ * IMPORTANT: the comparison values must be numeric values or at least be
+ * attempted to be converted to numeric values! If one of the arguments is
+ * something crazy like an empty string, unexpected behavior will occur because
+ * JavaScript.
+ *
+ * @param  {function} [objGetterFn] A function to obtain the comparison value
+ *  from the comparator function arguments
+ */
+function makeNumericValueComparator (objGetterFn) {
+  return (a, b) => {
+    let aVal
+    let bVal
+    if (typeof objGetterFn === 'function') {
+      aVal = objGetterFn(a)
+      bVal = objGetterFn(b)
+    } else {
+      aVal = a
+      bVal = b
+    }
+    // if both values aren't valid numbers, use the next sort criteria
+    if (isNaN(aVal) && isNaN(bVal)) return 0
+    // b is a valid number, b gets priority
+    if (isNaN(aVal)) return 1
+    // a is a valid number, a gets priority
+    if (isNaN(bVal)) return -1
```

Comment:
I'm a little worried about embedding this priority for numbers over strings in this comparator. If we decide that strings should come before numbers, we would have to refactor this function pretty heavily. There are a number of cases where we may actually want to give strings priority over ints (think about how TriMet [lists their routes](https://newplanner.trimet.org/map/#/route)). Perhaps this function should also except a boolean arg `stringsFirst` to indicate whether strings should be given higher priority in the sort (and these lines could use that arg)?

Reply:
I went ahead and added an option to forgo sorting by integer shortName. I think it'd be best to keep that because I can see it being quite common to have routes with "numeric" shortNames (ex: "3", "10", "20") and if we didn't sort by integer shortName, we'd get non-numerically sorted values (ex: "10", "20", "3").

Reply:
Sorry, I think maybe my initial point wasn't super clear. I do think we should sort shortNames as integers; however, I think that in some cases (maybe all) it might be preferable to have non-integer shortNames (and perhaps longNames) show up before integer shortNames in a list. E.g., Agency X may have numerically-named bus routes and color-named train lines. Because the train lines are more heavy utilized (and likely a shorter list) they might should show up first for better discoverability. So the order we're looking for would be:
```
BLUE
GREEN
RED
1
2
10
19
22
25
100
etc.
```

Reply:
I think so too. I'm also in favor of having one snapshot that has all the cases, so it shows what a normal route list could look like.

Reply:
@landonreed and @binh-dam-ibigroup, I just pushed a new commit that introduces two more comparator functions which I think will address these concerns. I've added a route type comparator that sorts based on route type. This will prioritize fixed guideway transit and ferries over buses. Secondly, I added a comparator that prioritizes routes with shortNames that begin with an alphabetic character.